### PR TITLE
Allow Dismissible to have a null direction

### LINKED ISSUE #24713: Add an option to set a Dismissble as 'active'
Attempt to disable the Dismissible() behavior by setting `.direction = null` causes assert at https://github.com/flutter/flutter/blob/6d134e0c866ce90a4dc7a02af6a8f1079c2a4fdc/packages/flutter/lib/src/widgets/dismissible.dart#L337

This is a feature enhancement to simplify customer code, not a code fault.

## Setup

Flutter version 0.11.9
Framework revision d48e6e433c (5 days ago), 2018-11-20 22:05:23 -0500
Engine revision 5c8147450d
Dart version 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)
Platform android-27, build-tools 27.0.3

## Repo Scenario

* ListView containing many Dismissible()
* Dismissible's contain a card
* Sometimes, all these cards can not be dismissed. There should be no sliding possible
* Sometimes, all these cards can be dismissed. Slide to remove them.

It would be desired to use the `Dismissible.direction` field to set the means to be dismissed. With flutter v0.11.9,  we can set that field to any choice of `DismissDirection` but unfortunately not to `null`. Naturally, `null` as a `.direction` would indicate there is no direction in which it can be dismissed...making the Dismissible a noop.

Support for `null` as a value to `Dismissible.direction` will simply customer code. Rather than using a condition expression and duplicating the child widget code, it can more easily be done by just setting the `direction` to null.

Today's unhappy duplicate code 😐

```dart
Listview.builder(...
  itemBuilder:
    return isDisabled ?
      CardWidgetAndCode(
        ...
      )
    : Dismissible(
      child: CardWidgetAndCode( //  <-- duplicate code and/or function call
        ...
      )
    )
)
```

Future happy simplified code 🙂
```dart
Listview.builder(...
  itemBuilder:
    Dismissible(
      direction: isDisabled ? null : DismissDirection.horizontal,
      child: CardWidgetAndCode(
        ...
      )
    )
)
```

## Log of assert

```
package:flutter/src/widgets/dismissible.dart': Failed assertion: line 337 pos 12: 'widget.direction != null': is not true
```

## Doctor

```
[√] Flutter (Channel beta, v0.11.9, on Microsoft Windows [Version 10.0.17763.165], locale en-US)
    • Flutter version 0.11.9 at C:\Users\dale\.flutter
    • Framework revision d48e6e433c (5 days ago), 2018-11-20 22:05:23 -0500
    • Engine revision 5c8147450d
    • Dart version 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)

[√] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at C:\Users\dale\.android\Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • ANDROID_HOME = C:\Users\dale\.android\Sdk
    • Java binary at: C:\Program Files\Android\android-studio-preview\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[√] Android Studio (version 3.2)
    • Android Studio at C:\Program Files\Android\android-studio-preview
    • Flutter plugin version 30.0.1
    • Dart plugin version 181.5656
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[√] VS Code, 64-bit edition (version 1.29.1)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 2.20.0

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 8.0.0 (API 26) (emulator)

• No issues found!
```

### PR DESCRIPTION
**Pull request by Sub6Resources**
## Description

This pull request allows `Dismissible.direction` to be `null`, which will help simplify user code when attempting to completely disable the dismiss gesture on a Dismissible:

Previous code:

```dart
Listview.builder(...
  itemBuilder:
    return isDisabled ?
      CardWidgetAndCode(
        ...
      )
    : Dismissible(
      child: CardWidgetAndCode( //  <-- duplicate code and/or function call
        ...
      )
    )
)
```

New simplified code:

```dart
Listview.builder(...
  itemBuilder:
    Dismissible(
      direction: isDisabled ? null : DismissDirection.horizontal,
      child: CardWidgetAndCode(
        ...
      )
    )
)
```

## Related Issues

Fixes #24713

## Tests

I added a test to ensure that attempting to dismiss a Dismissible with a null direction does not dismiss it.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/dismissible.dart [hunk 1]
```diff
@@ -509,6 +512,10 @@ class _DismissibleState extends State<Dismissible> with TickerProviderStateMixin
   Widget build(BuildContext context) {
     super.build(context); // See AutomaticKeepAliveClientMixin.
 
+    if (widget.direction == null) {
+      return widget.child;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/dismissible.dart]: If `widget.child` is `null`, `performRebuild` will throw. Could you assert in the constructor that `child` and `direction` can't both be `null`?

Comment (Sub6Resources) [reply packages/flutter/lib/src/widgets/dismissible.dart]: Ah, yeah. Good catch. Added an assertion to the constructor.

File: packages/flutter/test/widgets/dismissible_test.dart [hunk 2]
```diff
@@ -732,4 +732,16 @@ void main() {
     expect(dismissedItems, isEmpty);
     expect(confirmDismissDirection, DismissDirection.endToStart);
   });
+
```
Comment (Piinks) [inline packages/flutter/test/widgets/dismissible_test.dart]: Can you add tests that will trigger the conditions of your assertion to make sure it works as intended?

File: packages/flutter/lib/src/widgets/dismissible.dart [hunk 3]
```diff
@@ -97,6 +99,7 @@ class Dismissible extends StatefulWidget {
   }) : assert(key != null),
        assert(secondaryBackground == null || background != null),
        assert(dragStartBehavior != null),
+       assert(child != null || direction != null),
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/dismissible.dart]: Can we add more to the documentation, or an assertion message here that explains this?

File: packages/flutter/lib/src/widgets/dismissible.dart [hunk 4]
```diff
@@ -130,6 +133,8 @@ class Dismissible extends StatefulWidget {
   final DismissDirectionCallback onDismissed;
 
   /// The direction in which the widget can be dismissed.
+  ///
+  /// If null, the widget will not be dismissible.
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/dismissible.dart]: This would be a good spot to clarify that the child cannot be null if this is.

**Review by Piinks — COMMENTED** [review]
Hi @Sub6Resources,
Welcome! And thank you for the contribution! 🎉

**Comment by Sub6Resources** [comment]
Thanks for the review @Piinks. As you recommended, I've added a test and a bit more to the documentation.

**Comment by Piinks** [comment]
Hey @Sub6Resources, sorry for the delay, the `material` label took this off my radar. 😅 
This change looks good. Can you update your branch and resolve the merge conflict? I think this will be ready to land then.

**Comment by goderbauer** [comment]
I am closing this for now. @Piinks volunteered to take this over and finish it in a separate PR.

**Comment by Piinks** [comment]
I may be able to just resolve the merge conflict here.

**Comment by Piinks** [comment]
Or not. 😝  There are some versioning issues that will be easier to resolve by just opening a new PR. I'll link the two when it's up.

**Pull request closed without merging**